File: sdks/python/settlekit/resources/subscriptions.py
"""Subscriptions resource — create and manage recurring subscriptions.

Subscriptions require a recurring price (monthly/yearly). Creating one also
grants a subscription entitlement for the product; both are returned together.
"""

from __future__ import annotations

from typing import Any, Optional

from ._base import AsyncResource, SyncResource, clean_body, clean_params


def _create_body(
    *,
    organization_id: str,
    customer_id: str,
    product_id: str,
    price_id: str,
    cancel_at_period_end: Optional[bool],
) -> dict[str, Any]:
    return clean_body(
        {
            "organizationId": organization_id,
            "customerId": customer_id,
            "productId": product_id,
            "priceId": price_id,
            "cancelAtPeriodEnd": cancel_at_period_end,
        }
    )


class Subscriptions(SyncResource):
    """Synchronous subscriptions namespace."""

    def list(self, *, organization_id: Optional[str] = None) -> Any:
        """List subscriptions for an organization. ``GET /v1/subscriptions``."""
        return self._request(
            "GET",
            "/v1/subscriptions",
            params=clean_params({"organizationId": organization_id}),
        )

    def create(
        self,
        *,
        organization_id: str,
        customer_id: str,
        product_id: str,
        price_id: str,
        cancel_at_period_end: Optional[bool] = None,
    ) -> Any:
        """Create a subscription from a recurring price. ``POST /v1/subscriptions``."""
        return self._request(
            "POST",
            "/v1/subscriptions",
            body=_create_body(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
                price_id=price_id,
                cancel_at_period_end=cancel_at_period_end,
            ),
        )

    def get(self, subscription_id: str) -> Any:
        """Fetch a subscription by id. ``GET /v1/subscriptions/:id``."""
        return self._request("GET", f"/v1/subscriptions/{subscription_id}")

    def renew(self, subscription_id: str) -> Any:
        """Advance a subscription to its next period. ``POST /v1/subscriptions/:id/renew``."""
        return self._request("POST", f"/v1/subscriptions/{subscription_id}/renew")

    def cancel(self, subscription_id: str) -> Any:
        """Cancel a subscription. ``POST /v1/subscriptions/:id/cancel``."""
        return self._request("POST", f"/v1/subscriptions/{subscription_id}/cancel")


class AsyncSubscriptions(AsyncResource):
    """Asynchronous subscriptions namespace."""

    async def list(self, *, organization_id: Optional[str] = None) -> Any:
        """List subscriptions for an organization. ``GET /v1/subscriptions``."""
        return await self._request(
            "GET",
            "/v1/subscriptions",
            params=clean_params({"organizationId": organization_id}),
        )

    async def create(
        self,
        *,
        organization_id: str,
        customer_id: str,
        product_id: str,
        price_id: str,
        cancel_at_period_end: Optional[bool] = None,
    ) -> Any:
        """Create a subscription from a recurring price. ``POST /v1/subscriptions``."""
        return await self._request(
            "POST",
            "/v1/subscriptions",
            body=_create_body(
                organization_id=organization_id,
                customer_id=customer_id,
                product_id=product_id,
                price_id=price_id,
                cancel_at_period_end=cancel_at_period_end,
            ),
        )

    async def get(self, subscription_id: str) -> Any:
        """Fetch a subscription by id. ``GET /v1/subscriptions/:id``."""
        return await self._request("GET", f"/v1/subscriptions/{subscription_id}")

    async def renew(self, subscription_id: str) -> Any:
        """Advance a subscription to its next period. ``POST /v1/subscriptions/:id/renew``."""
        return await self._request(
            "POST", f"/v1/subscriptions/{subscription_id}/renew"
        )

    async def cancel(self, subscription_id: str) -> Any:
        """Cancel a subscription. ``POST /v1/subscriptions/:id/cancel``."""
        return await self._request(
            "POST", f"/v1/subscriptions/{subscription_id}/cancel"
        )

File: apps/cli/src/commands/subscriptions.ts
/**
 * `settlekit subscriptions` — manage recurring subscriptions.
 *
 *   list                 GET  /v1/subscriptions
 *   create               POST /v1/subscriptions
 *   get <id>             GET  /v1/subscriptions/:id
 *   renew <id>          POST /v1/subscriptions/:id/renew
 *   cancel <id>         POST /v1/subscriptions/:id/cancel
 *
 * Subscriptions require a recurring price (monthly/yearly). Creating one also
 * grants a subscription entitlement for the product.
 */
import type { Command } from "commander";
import { buildContext } from "../context.js";
import type { Money } from "../api.js";

interface Subscription extends Record<string, unknown> {
  id: string;
  customerId: string;
  productId: string;
  status: string;
  currentPeriodEnd?: string;
  amount?: Money;
}

interface CreateResult {
  subscription: Subscription;
  entitlement?: Record<string, unknown>;
}

const SUBSCRIPTION_COLUMNS = [
  { header: "ID", value: (s: Subscription) => s.id },
  { header: "CUSTOMER", value: (s: Subscription) => s.customerId },
  { header: "PRODUCT", value: (s: Subscription) => s.productId },
  { header: "STATUS", value: (s: Subscription) => s.status },
  { header: "RENEWS", value: (s: Subscription) => s.currentPeriodEnd },
] as const;

export function registerSubscriptions(program: Command): void {
  const subscriptions = program
    .command("subscriptions")
    .description("Manage recurring subscriptions");

  subscriptions
    .command("list")
    .description("List subscriptions for an organization")
    .option("--organization-id <id>", "Organization id (defaults to platform org)")
    .action(async function (this: Command) {
      const opts = this.opts();
      const ctx = buildContext(this);
      const rows = await ctx.client.get<Subscription[]>("/v1/subscriptions", {
        organizationId: opts.organizationId,
      });
      ctx.printList(rows, SUBSCRIPTION_COLUMNS);
    });

  subscriptions
    .command("create")
    .description("Create a subscription from a recurring price")
    .requiredOption("--organization-id <id>", "Organization id")
    .requiredOption("--customer-id <id>", "Customer id")
    .requiredOption("--product-id <id>", "Product id")
    .requiredOption("--price-id <id>", "Recurring price id (monthly/yearly)")
    .option("--cancel-at-period-end", "Schedule cancellation at the end of the first period", false)
    .action(async function (this: Command) {
      const opts = this.opts();
      const ctx = buildContext(this);
      const body: Record<string, unknown> = {
        organizationId: opts.organizationId,
        customerId: opts.customerId,
        productId: opts.productId,
        priceId: opts.priceId,
      };
      if (opts.cancelAtPeriodEnd) body.cancelAtPeriodEnd = true;
      const result = await ctx.client.post<CreateResult>("/v1/subscriptions", body);
      ctx.printRecord(result.subscription);
    });

  subscriptions
    .command("get <id>")
    .description("Fetch a single subscription")
    .action(async function (this: Command, id: string) {
      const ctx = buildContext(this);
      const sub = await ctx.client.get<Subscription>(
        `/v1/subscriptions/${encodeURIComponent(id)}`,
      );
      ctx.printRecord(sub);
    });

  subscriptions
    .command("renew <id>")
    .description("Advance a subscription to its next billing period")
    .action(async function (this: Command, id: string) {
      const ctx = buildContext(this);
      const sub = await ctx.client.post<Subscription>(
        `/v1/subscriptions/${encodeURIComponent(id)}/renew`,
      );
      ctx.printRecord(sub);
    });

  subscriptions
    .command("cancel <id>")
    .description("Cancel a subscription")
    .action(async function (this: Command, id: string) {
      const ctx = buildContext(this);
      const sub = await ctx.client.post<Subscription>(
        `/v1/subscriptions/${encodeURIComponent(id)}/cancel`,
      );
      ctx.printRecord(sub);
    });
}
